python/pyarrow/tests/test_udf.py

```diff
@@ -0,0 +1,350 @@
+            pa.array([5, 10], pa.int64())
+        ],
+        [
+            pa.array([2, 3], pa.int64()),
+            pa.array([10, 20], pa.int64()),
+            pa.array([3, 7], pa.int64()),
+            pa.array([20, 30], pa.int64()),
+            pa.array([5, 10], pa.int64())
+        ]
+    ]
+
+
+@pytest.fixture
+def expected_outputs():
+    return [
+        # scalar output data
+        pa.scalar(11, pa.int64()),  # 10 + 1
+        pa.scalar(20, pa.int64()),  # 10 * 2
+        pa.scalar(25, pa.int64()),  # 10 * 2 + 5
+        pa.scalar(85, pa.int64()),  # (2 * 10) + (3 * 20) + 5
+        # array output data
+        pa.array([11, 21], pa.int64()),  # [10 + 1, 20 + 1]
+        pa.array([20, 80], pa.int64()),  # [10 * 2, 20 * 4]
+        pa.array([25, 90], pa.int64()),  # [(10 * 2) + 5, (20 * 4) + 10]
+        # [(2 * 10) + (3 * 20) + 5, (3 * 20) + (7 * 30) + 10]
+        pa.array([85, 280], pa.int64())
+    ]
+
+
+def test_scalar_udf_function_with_scalar_data(function_names,
+                                              function_arities,
+                                              function_input_types,
+                                              function_output_types,
+                                              function_docs,
+                                              functions,
+                                              function_inputs,
+                                              expected_outputs):
+
+    # Note: 2 * -> used to duplicate the list
+    # Because the values are same irrespective of the type i.e scalar or array
+    for name, \
+        arity, \
+        in_types, \
+        out_type, \
+        doc, \
+        function, \
+        input, \
+        expected_output in zip(function_names,
+                               2 * function_arities,
+                               function_input_types,
+                               2 * function_output_types,
+                               2 * function_docs,
+                               2 * functions,
+                               function_inputs,
+                               expected_outputs):
+
+        register_function(name, arity, doc, in_types, out_type, function)
+
+        func = pc.get_function(name)
+        assert func.name == name
+
+        result = pc.call_function(name, input)
+        assert result == expected_output
+
+
+def test_udf_input():
+    def unary_scalar_function(scalar):
+        return pc.call_function("add", [scalar, 1])
+
+    # validate arity
+    arity = -1
+    func_name = "py_scalar_add_func"
+    in_types = [InputType.scalar(pa.int64())]
+    out_type = pa.int64()
+    doc = get_function_doc("scalar add function", "scalar add function",
+                           ["scalar_value"])
+    try:
+        register_function(func_name, arity, doc, in_types,
+                          out_type, unary_scalar_function)
+    except Exception as ex:
+        assert isinstance(ex, AssertionError)
```

Comment:
use `pytest.raises`, and the proper exception for most of these is TypeError, and Python should raise it automatically as mentioned above